packages/jest-haste-map/src/index.js

```diff
@@ -0,0 +1,370 @@
+ * type ModuleMap = {[platform: string]: ModuleMetaData};
+ *
+ * //
+ * type ModuleMetaData = {
+ *   path: string, // the path to look up the file object in `files`.
+ *   type: string, // the module type (either `package` or `module`).
+ * };
+ *
+ * Note that the data structures described above are conceptual only. The actual
+ * implementation uses arrays and constant keys for metadata storage. Instead of
+ * `{id: 'flatMap', mtime: 3421, visited: true, dependencies: []}` the real
+ * representation is similar to `['flatMap', 3421, 1, []]` to save storage space
+ * and reduce parse and write time of a big JSON blob.
+ *
+ * The HasteMap is created as follows:
+ *  1. read data from the cache or create an empty structure.
+ *  2. crawl the file system.
+ *     * empty cache: crawl the entire file system.
+ *     * cache available:
+ *       * if watchman is available: get file system delta changes.
+ *       * if watchman is unavailable: crawl the entire file system.
+ *     * build metadata objects for every file. This builds the `files` part of
+ *       the `HasteMap`.
+ *  3. parse and extract metadata from changed files.
+ *     * this is done in parallel over worker processes to improve performance.
+ *     * the worst case is to parse all files.
+ *     * the best case is no file system access and retrieving all data from
+ *       the cache.
+ *    * the average case is a small number of changed files.
+ *  4. serialize the new `HasteMap` in a cache file.
+ *     Worker processes can directly access the cache through `HasteMap.read()`.
+ *
+ */
+class HasteMap {
+
+  constructor(options) {
+    this._options = {
+      cacheDirectory: options.cacheDirectory || os.tmpDir(),
+      extensions: options.extensions,
+      ignorePattern: options.ignorePattern,
+      maxWorkers: options.maxWorkers,
+      mocksPattern:
+        options.mocksPattern ? new RegExp(options.mocksPattern) : null,
+      name: options.name,
+      platforms: options.platforms,
+      resetCache: options.resetCache,
+      roots: options.roots,
+      useWatchman:
+        options.useWatchman === undefined ? true : options.useWatchman,
+    };
+
+    const list = options.providesModuleNodeModules;
+    this._whitelist = (list && list.length)
+      ? new RegExp('(' + NODE_MODULES + '(?:' + list.join('|') + '))', 'g')
+      : null;
+
+    this._cachePath = HasteMap.getCacheFilePath(
+      this._options.cacheDirectory,
+      VERSION,
+      this._options.name,
+      this._options.roots.join(':'),
+      this._options.extensions.join(':'),
+      this._options.platforms.join(':'),
+      options.mocksPattern
+    );
+    this._buildPromise = null;
+    this._workerPromise = null;
+    this._workerFarm = null;
+  }
+
+  static getCacheFilePath(tmpdir) {
+    const hash = crypto.createHash('md5');
+    Array.from(arguments).slice(1).forEach(arg => hash.update(arg));
+    return path.join(tmpdir, hash.digest('hex'));
```

Comment:
good idea! Let's use the name from the options and make it `tmpdir/<name>-hash`.